### PR TITLE
[ML] Reuse SourceDestValidator for data frame analytics

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit removes validation logic of source and dest indices
for data frame analytics and replaces it with using the common
`SourceDestValidator` class which is already used by transforms.
This way the validations and their messages become consistent
while we reduce code.

This means that where these validations fail the error messages
will be slightly different for data frame analytics.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java [hunk 1]
```diff
@@ -46,7 +46,6 @@
 
     // messages
     public static final String SOURCE_INDEX_MISSING = "Source index [{0}] does not exist";
-    public static final String SOURCE_LOWERCASE = "Source index [{0}] must be lowercase";
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java]: This was unused.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java [hunk 2]
```diff
@@ -427,13 +417,13 @@ public void validate(Context context, ActionListener<Context> listener) {
                 return;
             }
 
-            if (context.resolvedSource.contains(destIndex)) {
+            if (context.resolveSource().contains(destIndex)) {
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java]: nice spot!

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java [hunk 3]
```diff
@@ -260,17 +274,17 @@ public void testCheck_GivenMixedMissingAndExistingConcreteSourceIndex() throws I
...
+//            listener -> simpleNonRemoteValidator.validate(
+//                CLUSTER_STATE,
+//                new String[] { SOURCE_1, "missing" },
+//                "dest",
+//                Collections.emptyList(),
+//                listener
+//            ),
+//            true,
+//            null
+//        );
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java]: ^ needs re-activation or should be removed

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java [hunk 4]
```diff
@@ -464,21 +478,11 @@ public void testCheck_GivenDestIndexIsAliasThatMatchesMultipleIndices() throws I
                 CLUSTER_STATE,
                 new String[] { SOURCE_1 },
                 DEST_ALIAS,
-                SourceDestValidator.NON_DEFERABLE_VALIDATIONS,
+                Collections.emptyList(),
```
Comment (hendrikmuhs) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java]: not getting this change, should the test be removed?

**Review by hendrikmuhs — APPROVED** [review]
LGTM, some re-factoring leftovers which only need some cleanup (delete)

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java (hunk 4 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java]: These second assertions in those tests seem to be testing that when the validation is not present then the error does not occur. Previously the `NON_DEFERABLE_VALIDATIONS` set was passed around but I could not see what the difference was compared to passing an empty list. Does this make sense?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java (hunk 3 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidatorTests.java]: Reactivated it. Good catch.

**Pull request merged**